### PR TITLE
More simplifications

### PR DESCRIPTION
**Pull request by hermanventer**
Do more simplification inside refineWithPathCondition. (Right now, it is a bit awkward to use the actual simplifier here.)

Also tweak various things and delete some code that is a) uncovered by tests and b) probably not reachable anyway.

With these changes, Prepack can now deal with all the modules that are conditionally required in our benchmark application.

### CODE REVIEW

File: src/evaluators/IfStatement.js [hunk 1]
```diff
@@ -30,6 +31,7 @@ export function evaluate(
   let exprRef = env.evaluate(ast.test, strictCode);
   // 2. Let exprValue be ToBoolean(? GetValue(exprRef))
   let exprValue = GetValue(realm, exprRef);
+  if (exprValue instanceof AbstractValue) exprValue = simplifyAbstractValue(realm, exprValue);
```
Comment (NTillmann) [inline src/evaluators/IfStatement.js]: Is there an unwritten rule now that all calls to `GetValue` should be followed by a call to `simplifyAbstractValue`?

File: src/methods/join.js [hunk 2]
```diff
@@ -443,7 +443,7 @@ export function joinMaps<K, V>(
 // where the join is defined to be just m1[key] if m1[key] === m2[key] and
 // and abstract value with expression "joinCondition ? m1[key] : m2[key]" if not.
 export function joinBindings(realm: Realm, joinCondition: AbstractValue, m1: Bindings, m2: Bindings): Bindings {
-  function getAbstractValue(v1: void | Value, v2: void | Value): AbstractValue {
+  function getAbstractValue(v1: void | Value, v2: void | Value): Value {
```
Comment (NTillmann) [inline src/methods/join.js]: Why the change of return type? To make Flow happy? Don't see any other dataflow reason.

File: src/methods/join.js [hunk 3]
```diff
@@ -480,10 +480,12 @@ export function joinValuesAsConditional(
   condition: AbstractValue,
   v1: void | Value,
   v2: void | Value
-): AbstractValue {
+): Value {
   let result = AbstractValue.createFromConditionalOp(realm, condition, v1, v2);
-  if (v1) result.mightBeEmpty = v1.mightHaveBeenDeleted();
-  if (v2 && !result.mightBeEmpty) result.mightBeEmpty = v2.mightHaveBeenDeleted();
+  if (result instanceof AbstractValue) {
```
Comment (NTillmann) [inline src/methods/join.js]: Interesting and a bit surprising to see this treatment of the empty value. So our "conditional" is really something different than just the ternary JavaScript operator. I didn't realize that.

File: src/utils/paths.js [hunk 4]
```diff
@@ -53,27 +60,44 @@ function pushPathCondition(condition: Value) {
     pushPathCondition(left);
     pushPathCondition(right);
   } else {
+    if (condition.kind === "!=" || condition.kind === "==") {
```
Comment (NTillmann) [inline src/utils/paths.js]: The increasing asymmetry of `pushPathCondition` vs `pushInversePathCondition` worries me.

File: src/values/AbstractValue.js [hunk 5]
```diff
@@ -316,36 +290,74 @@ export default class AbstractValue extends Value {
...
+      if (op === "&&" || op === "||") {
+        let [left, right] = this.args;
+        invariant(left instanceof AbstractValue); // otherwise the factory would have simplified it.
+        let refinedLeft = left.refineWithPathCondition();
+        let refinedRight = right instanceof AbstractValue ? right.refineWithPathCondition() : right;
+        // todo: remove this check when there is an alternative way to indicate that an intrinsic object is nullable
+        if (!(refinedLeft instanceof AbstractObjectValue && refinedLeft.isIntrinsic())) {
+          // true && y <=> y
+          // true || y <=> true
+          if (!refinedLeft.mightNotBeTrue()) return op === "&&" ? refinedRight : realm.intrinsics.true;
```
Comment (cblappert) [inline src/values/AbstractValue.js]: Why is it `realm.intrinsics.true` not `refinedLeft`?

File: src/methods/join.js (hunk 2 above)
Comment (hermanventer) [reply src/methods/join.js]: Join calls AbstractValue. createFromConditionalOp which now can return a concrete value as a result of refinement and simplification.

File: test/serializer/abstract/WaitGenerator4.js [hunk 6]
```diff
@@ -1,6 +1,6 @@
 (function() {
     let x = global.__abstract ? __abstract("boolean", "(true)") : true;
-    let y = global.__abstract ? __abstract("boolean", "(true)") : true;
+    let y = global.__abstract ? __abstract("boolean", "((true))") : true;
```
[the remark this answers is not in the document]
Comment (hermanventer) [reply test/serializer/abstract/WaitGenerator4.js]: Two abstract values with the same type and intrinsic name will be treated as identical during simplification. Having just one value reach the serializer breaks the test case.

File: src/utils/paths.js (hunk 4 above)
Comment (hermanventer) [reply src/utils/paths.js]: Good point. There is not reason why pushInversePathCondition should not also handle != and ==.

File: src/values/AbstractValue.js [hunk 7]
```diff
@@ -316,36 +290,74 @@ export default class AbstractValue extends Value {
...
-    let result = this;
-    for (let pathCondition of this.$Realm.pathConditions) {
-      if (pathCondition.implies(condition)) {
-        result = trueVal;
-        break;
+
+    let realm = this.$Realm;
+    if (realm.pathConditions.length === 0) return this;
+    let op = this.kind;
+    let result = (() => {
```
[the remark this answers is not in the document]
Comment (hermanventer) [reply src/values/AbstractValue.js]: It is so that we can use return inside this block, but then also post process the return value before actually returning it.

File: src/evaluators/IfStatement.js (hunk 1 above)
Comment (hermanventer) [reply src/evaluators/IfStatement.js]: This commit belongs to a previous pull request. Good question. Right now, things happen here because I did not want to add simplifier.js to the long cycle. I'm not happy about this and have some ideas of how to refactor things a bit more nicely. I'll add some relevant todo comments in a future pull request.

File: src/methods/join.js (hunk 3 above)
Comment (hermanventer) [reply src/methods/join.js]: createFromConditionalOp can now return concrete values and concrete values do not have the mightBeEmpty property. I introduced mightBeEmpty a long time ago to make it easier to model joins between branches where a property is missing and branches where it is present. We could possibly do the same for undefined and null, but since empty is not really a value, while the others are the analogy is somewhat strained. I'm still in a muddle about the whole thing and I'm putting it off until later.

File: src/values/AbstractValue.js (hunk 5 above)
Comment (hermanventer) [reply src/values/AbstractValue.js]: It should be refinedLeft.

**Comment by hermanventer** [comment]
Made changes to address review feedback.

File: test/serializer/abstract/WaitGenerator4.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (yinghuitan) [reply test/serializer/abstract/WaitGenerator4.js]: Can you add this as a comment into the testcase? People read this testcase will ask frequently.

File: src/values/AbstractValue.js [hunk 8]
```diff
@@ -510,7 +492,25 @@ export default class AbstractValue extends Value {
...
+  ): Value {
+    // c ? x : x <=> x
+    if (left !== undefined && right !== undefined && left.equals(right)) return left;
+    // x ? x : y <=> x || y
+    if (left instanceof AbstractValue && condition.equals(left) && right)
+      return AbstractValue.createFromLogicalOp(realm, "||", left, right, loc);
+    // y ? x : y <=> y && x
+    if (right instanceof AbstractValue && condition.equals(right) && left)
+      return AbstractValue.createFromLogicalOp(realm, "&&", right, left, loc);
+    // c ? (c ? x : y) : z <=> c ? x : z
```
Comment (yinghuitan) [inline src/values/AbstractValue.js]: I am a bit vague on this. Is it safe to do this optimization while "y" has some side-effect? For example:
```js
var y = sideEffect();
var result = c ? (c ? x : y) : z;
```

File: src/values/AbstractValue.js [hunk 9]
```diff
@@ -348,18 +305,39 @@ export default class AbstractValue extends Value {
...
+          if (!refinedLeft.mightNotBeTrue()) return op === "&&" ? refinedRight : refinedLeft;
+          // (x == false) && y <=> x
+          // false || y <=> y
+          if (!refinedLeft.mightNotBeFalse()) return op === "||" ? refinedRight : refinedLeft;
+        }
+        if (refinedLeft.getType() === BooleanValue && refinedRight.getType() === BooleanValue) {
+          // (x: boolean) && true <=> x
+          // x || true <=> true
+          if (!refinedRight.mightNotBeTrue()) return op === "&&" ? refinedLeft : realm.intrinsics.true;
+          // (x: boolean) && false <=> false
```
Comment (yinghuitan) [inline src/values/AbstractValue.js]: Missing a comment for "||"

Comment (yinghuitan) [inline src/values/AbstractValue.js]: Similar to my comment below, what if "x" has side-effect? I do not think it is safe to optimize it away. Or maybe there is a all-arguments-non-side-effect pre-requirement checking somewhere I missed?

File: src/values/AbstractValue.js [hunk 10]
```diff
@@ -490,6 +464,14 @@ export default class AbstractValue extends Value {
       rightValues = new ValuesDomain(right);
     }
 
+    if (left.getType() === BooleanValue && right.getType() === BooleanValue) {
+      // (x: boolean) && true <=> x
+      // x || true <=> true
+      if (!right.mightNotBeTrue()) return op === "&&" ? left : realm.intrinsics.true;
+      // (x: boolean) && false <=> false
```
Comment (yinghuitan) [inline src/values/AbstractValue.js]: Missing comment for "||"

File: src/values/AbstractValue.js [hunk 11]
```diff
@@ -490,6 +464,14 @@ export default class AbstractValue extends Value {
       rightValues = new ValuesDomain(right);
     }
 
+    if (left.getType() === BooleanValue && right.getType() === BooleanValue) {
```
Comment (yinghuitan) [inline src/values/AbstractValue.js]: This section is the same logic as the block starting from line325. Maybe fold into a helper function to avoid duplication?

Comment (yinghuitan) [reply src/values/AbstractValue.js]: Also, any reason, the code here only performs this one kind of refinement, while the code you added in refineWithPathCondition() does several more?

File: src/values/AbstractValue.js (hunk 8 above)
Comment (NTillmann) [reply src/values/AbstractValue.js]: Arguments to abstract values must always be pure, so in that sense, it's safe.

File: src/values/AbstractValue.js (hunk 9 above)
Comment (NTillmann) [reply src/values/AbstractValue.js]: It's by construction. Difficult to check/enforce, but very fundamental to the nature of abstract values.

File: src/values/AbstractValue.js [hunk 12]
```diff
@@ -348,18 +305,39 @@ export default class AbstractValue extends Value {
     }
 
     let realm = this.$Realm;
+    if (realm.pathConditions.length === 0) return this;
```
Comment (NTillmann) [inline src/values/AbstractValue.js]: Why? Is this a performance optimization?

File: src/values/AbstractValue.js [hunk 13]
```diff
@@ -348,18 +305,39 @@ export default class AbstractValue extends Value {
...
+          // true || y <=> true
+          if (!refinedLeft.mightNotBeTrue()) return op === "&&" ? refinedRight : refinedLeft;
+          // (x == false) && y <=> x
+          // false || y <=> y
+          if (!refinedLeft.mightNotBeFalse()) return op === "||" ? refinedRight : refinedLeft;
+        }
+        if (refinedLeft.getType() === BooleanValue && refinedRight.getType() === BooleanValue) {
+          // (x: boolean) && true <=> x
+          // x || true <=> true
+          if (!refinedRight.mightNotBeTrue()) return op === "&&" ? refinedLeft : realm.intrinsics.true;
```
Comment (NTillmann) [inline src/values/AbstractValue.js]: After refining the operands, the following simplification rules are generally applicable, and have nothing to do path conditions. Why are they here and not in a more general places?

File: src/values/AbstractValue.js (hunk 11 above)
Comment (NTillmann) [reply src/values/AbstractValue.js]: Good point. I don't think there should be a helper function for this. There should be just one way of constructing a logical abstract value, and it should always do the optimization.

File: src/values/AbstractValue.js (hunk 12 above)
Comment (hermanventer) [reply src/values/AbstractValue.js]: Yes, but mostly it makes debugging less painful.

**Comment by hermanventer** [comment]
I agree that simplification should happen in one place only. I have some ideas on that and will follow up later.

File: src/values/AbstractValue.js (hunk 11 above)
Comment (hermanventer) [reply src/values/AbstractValue.js]: Agreed.

File: test/serializer/abstract/WaitGenerator4.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (hermanventer) [reply test/serializer/abstract/WaitGenerator4.js]: sure.

File: src/values/AbstractValue.js (hunk 9 above)
Comment (hermanventer) [reply src/values/AbstractValue.js]: will add

File: src/values/AbstractValue.js (hunk 10 above)
Comment (hermanventer) [reply src/values/AbstractValue.js]: will add.

File: src/values/AbstractValue.js (hunk 13 above)
Comment (hermanventer) [reply src/values/AbstractValue.js]: Agreed. Will attempt it in a future diff.

**Comment by hermanventer** [comment]
Made requested changes.
